### PR TITLE
Improved git status

### PR DESCRIPTION
**Pull request by BenoitZugmeyer**
Improved git status, based on [the one I'm using currently](https://github.com/BenoitZugmeyer/dotfiles/blob/master/home/.zsh/conf/prompt-git-info.zsh). This is a bit opinionated, so feel free to tell me what I should change (or make adjustements yourself).

* show the commit short id instead of 'HEAD' if no shorthand is found
* show the number of commits behind or ahead of the upstream branch
* show the number of files:
  * changed in the index
  * changed in the work tree
  * untracked
  * in conflict
* show the current action (rebase, merge, ...)

### CODE REVIEW

**Comment by xcambar** [comment]
Your PR "forces me" to introduce an idea I had for later: Themes and Segments.

I'd very much like to let the user choose between what they want:
*  is currently implemented (_aka_ `my_branch*` format) 
* your version

That would be done via command-line arguments. I thought of `--git-short` and `--git-detailed` or something along these lines.

The main point is that I won't use your prompt, but I think yours has something to offer, so I think it's silly from me to close the PR and call it a day.

so, tl;dr; update your PR so it is handled with command-line arguments, and I'll happily review it. I can give it a try as well on another PR on top of master to pave the way...

What do you think?

**Comment by BenoitZugmeyer** [comment]
I understand your concern and since you ask me: if I was the maintainer of purs, I'd either chose to have no option (no more option than the current implementation, very opinionated) or fully configurable with a template, something like:
```
{cyan}{git_head_name}{/cyan}
{green}{git_clean_status}{/green}
{cyan}{git_commits_ahead "↑"}{git_commits_behind "↓"}{/cyan}
```

But I can add a `--git-detailed` option, it's not a problem!

**Comment by xcambar** [comment]
Incredible: Using a template was in the back of my mind from the beginning!
I didn't mention it because I didn't want this PR to hold on too long nor inflate too much before being merged.

Asking for the option as you implemented it was all about finding a meeting point halfway through.

**Comment by xcambar** [comment]
I should mention though that having the full-template solution requires more data points and raises many questions:
* how many segments can be thought of (<=> is it worth it)?
* What is the cost in terms of performance (performance is more important than features in this project, for me)?
* Should they be packed in the app or come as plugin (I thought of using git commands syntax: `git-ls` => `git ls` as a basename for extensions)?
* as a fun architecture exercise: some plugins would be async (weather, git fetch...)? How to implement that? with a daemon?
* In the end... is it not just about reimplementing Powerline in Rust?

With that in mind, you'd understand better the dilemma of being very open VS being very opinionated I'm sure :D

**Comment by pquentin** [comment]
As a purs user, for me the order of importance is 1/ performance 2/ features 3/ opinionated. Of course you don't have to agree.

I still use part of my old right prompt (RPROMPT) to see the remote name and the number of stashes, but it's slow, so using fast templates would be nice.

**Comment by xcambar** [comment]
@pquentin that makes total sense, thanks!

@BenoitZugmeyer can you add some space between the elements of the output? It appears very stacked for me (another argument for templates? :) ) and we have room considering it's in the first line of the prompt.

FYI, I'll open another PR discussing the glyphs. I need to do some research, but they're not meaningful to me, yet I don't mean to put too much of my own (lack of) habit in this.

**Comment by xcambar** [comment]
I've been picky and that was not necessary.

Thanks for the PR!
